conda-store-server/conda_store_server/plugins/plugin_context.py

```diff
@@ -0,0 +1,74 @@
+# Copyright (c) conda-store development team. All rights reserved.
+# Use of this source code is governed by a BSD-style
+# license that can be found in the LICENSE file.
+
+import io
+import logging
+import subprocess
+import uuid
+
+
+class PluginContext:
+    """The plugin context provides some useful attributes to a hook.
+
+    This includes
+        * the variables: conda_store, log, stdout, stderr
+        * the functions: run_command, run
+    """
+
+    def __init__(
+        self, conda_store=None, stdout=None, stderr=None, log_level=logging.INFO
+    ):
+        if stdout is not None and stderr is None:
+            stderr = stdout
+
+        self.id = str(uuid.uuid4())
+        self.stdout = stdout if stdout is not None else io.StringIO()
+        self.stderr = stderr if stderr is not None else io.StringIO()
+        self.log = logging.getLogger(
+            f"conda_store_server.plugins.plugin_context.{self.id}"
+        )
+        self.log.propagate = False
+        self.log.addHandler(logging.StreamHandler(stream=self.stdout))
+        self.log.setLevel(log_level)
+        self.conda_store = conda_store
+
+    def run_command(self, command, redirect_stderr=True, **kwargs):
+        """Runs command and immediately writes to logs"""
+        self.log.info(f"Running command: {' '.join(command)}")
```

Comment:
```suggestion
        self.log.info("Running command: %s, command)
```

Not sure what's canon in conda-store but it's usually preferred to defer stringification to log-time. Also, for clarity with commands and spaces, let's leave it as a list.